src/main/java/io/jenkins/plugins/opentelemetry/servlet/OpenTelemetryServletFilter.java

```diff
@@ -0,0 +1,292 @@
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.api.trace.SpanBuilder;
+import io.opentelemetry.api.trace.Tracer;
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.semconv.trace.attributes.SemanticAttributes;
+import org.apache.commons.lang.StringUtils;
+
+import javax.annotation.Nullable;
+import javax.servlet.Filter;
+import javax.servlet.FilterChain;
+import javax.servlet.FilterConfig;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Objects;
+import java.util.StringTokenizer;
+import java.util.stream.Collectors;
+
+public class OpenTelemetryServletFilter implements Filter {
+    private final Tracer tracer;
+
+    public OpenTelemetryServletFilter(Tracer tracer) {
+        this.tracer = tracer;
+    }
+
+    @Override
+    public void doFilter(ServletRequest servletRequest, ServletResponse servletResponse, FilterChain filterChain) throws IOException, ServletException {
+
+        if (servletRequest instanceof HttpServletRequest) {
+            _doFilter((HttpServletRequest) servletRequest, (HttpServletResponse) servletResponse, filterChain);
+        } else {
+            filterChain.doFilter(servletRequest, servletResponse);
+        }
+    }
+
+    public void _doFilter(HttpServletRequest servletRequest, HttpServletResponse servletResponse, FilterChain filterChain) throws IOException, ServletException {
+
+        String pathInfo = servletRequest.getPathInfo();
+        List<String> pathInfoTokens = Collections.list(new StringTokenizer(pathInfo, "/")).stream()
+            .map(token -> (String) token)
+            .filter(t -> !t.isEmpty())
+            .collect(Collectors.toList());
+
+
+        if (pathInfoTokens.isEmpty()) {
+            pathInfoTokens = Collections.singletonList("");
+        }
+
+
+        String rootPath = pathInfoTokens.get(0);
+        // The matched route (path template).
+        String httpRoute;
+
+        if (rootPath.equals("static") || rootPath.equals("adjuncts") || rootPath.equals("scripts") || rootPath.equals("plugin")) {
+            // skip
+            filterChain.doFilter(servletRequest, servletResponse);
+        } else if (rootPath.equals("$stapler")) {
+            // TODO handle URL pattern /$stapler/bound/ec328aeb-26be-43da-94a3-59f2d683131c/news
+            filterChain.doFilter(servletRequest, servletResponse);
+        } else {
+            SpanBuilder spanBuilder;
+            if (rootPath.equals("job")) {
+                // e.g /job/my-war/job/master/lastBuild/console
+                // e.g /job/my-war/job/master/2/console
+                ParsedJobUrl parsedJobUrl = parseJobUrl(pathInfoTokens);
```

Comment:
This all seems like a maintenance burden. Did you consider using a supported API like https://javadoc.jenkins.io/component/stapler/org/kohsuke/stapler/StaplerRequest.html#getAncestors()?

Reply:
Thanks I got trapped. I started adding HTTP request tracing to understand how the progressive rendering of logs worked in Jenkins and ended up this insane if then else.

I'll add a comment in the code to use this Stapler API.

Reply:
> adding HTTP request tracing to understand how the progressive rendering of logs worked in Jenkins

BTW `-Dstapler.trace`, automatic with `mvn hpi:run`.